tsl/src/nodes/vector_agg/grouping_policy_hash.c

```diff
@@ -0,0 +1,735 @@
+/*
+ * This file and its contents are licensed under the Timescale License.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-TIMESCALE for a copy of the license.
+ */
+
+/*
+ * This grouping policy groups the rows using a hash table. Currently it only
+ * supports a single fixed-size by-value compressed column that fits into a Datum.
+ */
+
+#include <postgres.h>
+
+#include <executor/tuptable.h>
+#include <nodes/pg_list.h>
+
+#include "grouping_policy.h"
+
+#include "nodes/decompress_chunk/compressed_batch.h"
+#include "nodes/vector_agg/exec.h"
+
+#ifdef USE_FLOAT8_BYVAL
+#define DEBUG_LOG(MSG, ...) elog(DEBUG3, MSG, __VA_ARGS__)
+#else
+/*
+ * On 32-bit platforms we'd have to use the cross-platform int width printf
+ * specifiers which are really unreadable.
+ */
+#define DEBUG_LOG(...)
+#endif
+
+/*
+ * We can use crc32 as a hash function, it has bad properties but takes only one
+ * cycle, which is why it is sometimes used in the existing hash table
+ * implementations.
+ */
+#ifdef USE_SSE42_CRC32C
+#include <nmmintrin.h>
+static pg_attribute_always_inline uint64
+hash64(uint64 x)
+{
+	return _mm_crc32_u64(~0ULL, x);
+}
+
+#else
+/*
+ * When we don't have the crc32 instruction, use the SplitMix64 finalizer.
+ */
+static pg_attribute_always_inline uint64
+hash64(uint64 x)
+{
+	x ^= x >> 30;
+	x *= 0xbf58476d1ce4e5b9U;
+	x ^= x >> 27;
+	x *= 0x94d049bb133111ebU;
+	x ^= x >> 31;
+	return x;
+}
+#endif
+
+/*
+ * For the hash table, use the generic Datum key that is mapped to the aggregate
+ * state index.
+ */
+typedef struct
+{
+	Datum key;
+	uint32 status;
+	uint32 agg_state_index;
```

Comment:
What about having the Agg state in the hash table entry instead? What are the pros/cons of having it here compared to in an outside array?

Reply:
The hash table should be as small as possible, because we're accessing its memory randomly, and the speed of this depends on whether it fits into the caches.